File: leetcode/101-200/135-candy/candy.py
"""LeetCode #135: Candy — two greedy passes."""


def candy(ratings):
    n = len(ratings)
    if n == 0:
        return 0
    c = [1] * n
    for i in range(1, n):
        if ratings[i] > ratings[i - 1]:
            c[i] = c[i - 1] + 1
    for i in range(n - 2, -1, -1):
        if ratings[i] > ratings[i + 1] and c[i] <= c[i + 1]:
            c[i] = c[i + 1] + 1
    return sum(c)


def lcg(seed, n, mod):
    x = seed
    out = []
    for _ in range(n):
        x = (x * 1103515245 + 12345) % 2147483648
        out.append(x % mod)
    return out


def main():
    MOD = 1000000007
    sink = 0
    cases = [[1, 0, 2], [1, 2, 2], [1, 3, 2, 2, 1], [5, 4, 3, 2, 1], [1, 2, 3, 4, 5]]
    for t in range(6):
        cases.append(lcg(t + 1, 12 + t * 4, 5))
    for r in cases:
        v = candy(r)
        print(f"n={len(r)}: {v}")
        sink = (sink * 1000003 + v) % MOD
    print(f"sink: {sink}")


if __name__ == "__main__":
    main()

File: leetcode/101-200/135-candy/bench/candy.rs
// Benchmark harness for LeetCode #135 — Candy.
// Mirrors candy.kara algorithm-for-algorithm, including the explicit
// descending index loop for the right-to-left pass (not `.rev()`), so the
// generated loop shape matches.

fn candy(ratings: &[i64]) -> i64 {
    let n = ratings.len() as i64;
    if n == 0 {
        return 0;
    }
    let mut c: Vec<i64> = vec![1; n as usize];

    let mut i: i64 = 1;
    while i < n {
        if ratings[i as usize] > ratings[(i - 1) as usize] {
            c[i as usize] = c[(i - 1) as usize] + 1;
        }
        i += 1;
    }

    i = n - 2;
    while i >= 0 {
        if ratings[i as usize] > ratings[(i + 1) as usize]
            && c[i as usize] <= c[(i + 1) as usize]
        {
            c[i as usize] = c[(i + 1) as usize] + 1;
        }
        i -= 1;
    }

    let mut total: i64 = 0;
    for i in 0..n {
        total += c[i as usize];
    }
    total
}

fn lcg(seed: i64, n: i64, cap: i64) -> Vec<i64> {
    let mut out: Vec<i64> = Vec::new();
    let mut x = seed;
    for _ in 0..n {
        x = (x * 1103515245 + 12345) % 2147483648;
        let wd0 = x / 65536;
        x = (x * 1103515245 + 12345) % 2147483648;
        out.push((wd0 * 32768 + x / 65536) % cap);
    }
    out
}

fn main() {
    let np: i64 = 8;
    let n: i64 = 200000;
    let iters: i64 = 150;

    let mut arrays: Vec<Vec<i64>> = Vec::new();
    for j in 0..np {
        if j % 2 == 0 {
            arrays.push(lcg(j + 1, n, 4));
        } else {
            arrays.push(lcg(j + 1, n, 100000));
        }
    }

    let mut sink: i64 = 0;
    for it in 0..iters {
        let idx = ((it * 3) % np) as usize;
        sink = (sink + candy(&arrays[idx])) % 1000000007;
    }
    println!("{}", sink);
}
